### ISSUE TITLE
The caching mechanism introduced in v12 kills all performance

### ISSUE THREAD

**Issue by DamienCassou** [mantoni/eslint_d.js#198]
Hi,

my setup has become very slow rencently and I managed to find out it is because of #190. If I apply the patch below (which is a tentative of partially reverting the problematic commit and doesn't make sense on its own), the speed is back to normal. Here is how my IDE uses eslint_d:

```
$ cat /project/dialogs.js | /eslint_d.js/bin/eslint_d.js --no-color --no-ignore --stdin --stdin-filename /project/dialogs.js --report-unused-disable-directives
```

And here the (non-sense) patch that fixes the issue for me:

```patch
 lib/linter.js | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/linter.js b/lib/linter.js
index fbdedca..45475c3 100644
--- a/lib/linter.js
+++ b/lib/linter.js
@@ -280,7 +280,7 @@   callback(null, output);
 /*
  * The core_d service entry point.
  */
-exports.invoke = async function (cwd, args, text, hash, callback) {
+exports.invoke = async function (cwd, args, text, mtime, callback) {
   process.chdir(cwd);
 
   let eslint_path_arg;
@@ -300,7 +300,7 @@ exports.invoke = async function (cwd, args, text, hash, callback) {
   let cache = eslintCache.get(cwd);
   if (!cache) {
     cache = createCache(cwd, eslint_path_arg);
-  } else if (hash !== cache.hash) {
+  } else if (mtime > cache.last_run) {
     clearRequireCache(cwd);
     cache = createCache(cwd, eslint_path_arg);
   }
@@ -308,7 +308,7 @@     clearRequireCache(cwd);
     callback(null);
     return;
   }
-  cache.hash = hash;
+    cache.hash = Date.now();
 
   const options = cache.eslint.ESLint
     ? options_eslint
-- 
2.36.0
```

Another workaround is to stay at 11.1.1.

**Comment by mantoni** [comment]
Interesting. It would be great to know where the performance degradation actually comes from. `eslint_d v12.0.0` is still way faster than `eslint` for me:

```
❯ time ./node_modules/.bin/eslint index.js

________________________________________________________
Executed in  271,77 millis    fish           external
   usr time  280,41 millis    0,07 millis  280,34 millis
   sys time   37,96 millis    1,21 millis   36,74 millis

❯ time eslint_d index.js

________________________________________________________
Executed in   84,13 millis    fish           external
   usr time   56,27 millis   55,00 micros   56,22 millis
   sys time   11,25 millis  880,00 micros   10,37 millis
```

**Comment by DamienCassou** [comment]
I'm comparing eslint_d with and without the new caching mechanism, not eslint_d and eslint:

```
$ time eslint_d-with-new-caching ...
real    0m4.963s
user    0m0.042s
sys     0m0.010s

$ time eslint_d-without-new-caching ...
real    0m0.171s
user    0m0.067s
sys     0m0.026s
```

Before each experiment, I kill the eslint_d process to make sure I'm running the latest version and I launch eslint_d once so it gets a chance to cache. The long delay when using the new caching is after each save of my file so it is *really* annoying: waiting 5s (sometimes 8s) each time I save my file is a bit too much compared to the previous version.

**Comment by mantoni** [comment]
Oh man, thank you for opening this issue ❤️. It turns out that `core_d` is doing the wrong thing entirely. There are two issues with checking common package manager files:

The `core_d` server appends file contents to the hash as they arrive. File reads happen concurrently, so the order is random. This leads to different hashes on every other call, even though the files didn't change. But even worse: It checks the files in the directory where the service was initially started. So calls from other directories don't even check the correct files. This was already an issue with previous versions. The entire feature makes no sense this way.

I'll try to make time to address these issues, unless you'd like to chime in?

**Comment by mantoni** [comment]
This is fixed with `v12.2.0`.

Here is what happened to the file hashing feature:

- It was completely removed from `core_d` and is re-implemented in `eslint_d` directly.
- It now checks the correct files in the `cwd`.
- It's even faster now because it only checking files that existed at cache creation time.
- It doesn't have the race condition flaws that where discovered in this issue.

Should you still have performance problems, please feel free to open a new issue.

Thank you again for reporting and being helpful with tracing it down. ❤️

**Comment by DamienCassou** [comment]
Thanks to *you* for fixing this issue so quickly! And thank you for eslint_d that is core to my workflow.
